// ngx-scrolltop/esm2015/lib/ngx-scrolltop.core.service.js
import { DOCUMENT } from '@angular/common';
import { Inject, Injectable } from '@angular/core';
export class NgxScrollTopCoreService {
    constructor(document) {
        this.document = document;
        this.scrolledFromTop = false;
        this.isBrowser = typeof window !== 'undefined';
    }
    onWindowScroll(mode) {
        var _a, _b;
        let show = false;
        const position = ((_a = this.document.documentElement) === null || _a === void 0 ? void 0 : _a.scrollTop) || ((_b = this.document.scrollingElement) === null || _b === void 0 ? void 0 : _b.scrollTop);
        switch (mode) {
            case 'classic':
                show = this.classicMode(position);
                break;
            case 'smart':
                show = this.smartMode(position);
                break;
        }
        return show;
    }
    classicMode(position) {
        let show = false;
        if (this.isBrowser && position > window.innerHeight) {
            show = true;
        }
        else {
            show = false;
        }
        return show;
    }
    smartMode(position) {
        let show = false;
        if (position === 0) {
            show = false;
            this.scrolledFromTop = false;
        }
        if (this.scrolledFromTop && this.scrollOffset > position) {
            show = true;
        }
        if (this.isBrowser && position > window.innerHeight * 2) {
            this.scrolledFromTop = true;
            this.scrollOffset = position;
        }
        return show;
    }
    scrollToTop() {
        if (this.isBrowser) {
            window.scroll({ top: 0, left: 0, behavior: 'smooth' });
        }
    }
}
NgxScrollTopCoreService.decorators = [
    { type: Injectable }
];
NgxScrollTopCoreService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
];
