// ngx-scrolltop/esm2015/lib/ngx-scrolltop.directive.js
import { Directive, ElementRef, HostListener, Input } from '@angular/core';
import { NgxScrollTopCoreService } from './ngx-scrolltop.core.service';
export class NgxScrollTopDirective {
    constructor(el, core) {
        this.el = el;
        this.core = core;
        this.mode = 'classic';
        this.hideElement();
    }
    onWindowScroll() {
        if (this.core.onWindowScroll(this.mode)) {
            this.showElement();
        }
        else {
            this.hideElement();
        }
    }
    onClick() {
        this.scrollToTop();
    }
    hideElement() {
        this.el.nativeElement.style.display = 'none';
    }
    showElement() {
        this.el.nativeElement.style.display = '';
    }
    scrollToTop() {
        this.core.scrollToTop();
    }
}
NgxScrollTopDirective.decorators = [
    { type: Directive, args: [{
                selector: '[ngxScrollTop]',
            },] }
];
NgxScrollTopDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: NgxScrollTopCoreService }
];
NgxScrollTopDirective.propDecorators = {
    mode: [{ type: Input, args: ['ngxScrollTopMode',] }],
    onWindowScroll: [{ type: HostListener, args: ['window:scroll',] }],
    onClick: [{ type: HostListener, args: ['click',] }]
};
